contracts/v2/destributor/Destributor.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: UNLICENSED
+
+pragma solidity 0.8.4;
+
+import './Interfaces.sol';
+import './MerkleProof.sol';
+
+contract Destributor {
+  mapping (uint256 => bytes32) public merkleRoot;
+  mapping (uint256 => mapping (uint256 => uint256)) private claims;
+  mapping (uint256 => bool) private cancelled;
+
+  uint256 public distribution;
+
+  address public immutable token;
+  address public immutable admin;
+  bool public paused;
+
+  event Distribute(bytes32 merkleRoot, uint256 distribution);
+  event Claim(uint256 index, address owner, uint256 amount);
+
+  /// @param t Address of an ERC20 token
+  /// @param r Initial merkle root
+  constructor(address t, bytes32 r) {
+    admin = msg.sender;
+    token = t;
+    merkleRoot[0] = r;
+  }
+
+  /// @notice Generate a new distribution, cancelling the would be current one
+  /// @param f The address of the wallet containing tokens to distribute
+  /// @param t The address that will receive any currently remaining distributions (will normally be the same as from)
+  /// @param a The amount of tokens in the new distribution
+  /// @param r The merkle root associated with the new distribution
+  function distribute(address f, address t, uint256 a, bytes32 r) external authorized(admin) returns (bool) {
+    // remove curent token balance
+    Erc20 erc = Erc20(token);
+    uint256 balance = erc.balanceOf(address(this));
+    erc.transfer(t, balance);
+        
+    // transfer enough tokens for new distribution
+    erc.transferFrom(f, address(this), a);
+
+    // we are working with the distribution, eventually bumping it...
+    uint256 current = distribution;
```

Comment:
Why store distribution in memory instead of just referencing the public var / bumping at the end

Reply:
to be in line with the gas-opts changes we made in the protocol contracts